sourcepolicy/engine.go

```diff
@@ -0,0 +1,226 @@
+package sourcepolicy
+
+import (
+	"context"
+	"strings"
+
+	"github.com/moby/buildkit/solver/pb"
+	spb "github.com/moby/buildkit/sourcepolicy/pb"
+	"github.com/moby/buildkit/util/bklog"
+	"github.com/pkg/errors"
+)
+
+var (
+	// ErrSourceDenied is returned by the policy engine when a source is denied by the policy.
+	ErrSourceDenied = errors.New("source denied by policy")
+
+	// ErrTooManyOps is returned by the policy engine when there are too many converts for a single source op.
+	ErrTooManyOps = errors.New("too many operations")
+)
+
+// Engine is the source policy engine.
+// It is responsible for evaluating a source policy against a source operation.
+// Create one with `NewEngine`
+//
+// Rule matching is delegated to the `Matcher` interface.
+// Mutations are delegated to the `Mutater` interface.
+type Engine struct {
+	pol     []*spb.Policy
+	matcher Matcher
```

Comment:
Do you have some plans with the `Matcher/Mutater` types/parameters or why the extra abstraction?

Reply:
I don't. I had these when I was first building it and wasn't sure what it would end up like.
I can remove the extra abstractions.